### PR TITLE
chore: useswapper cleanup

### PR DESCRIPTION
**Pull request by elmutt**
## Description

Cleanup useSwapper useages in web

- Swappers are now assumed to always use sellAmount -- simplifies logic in many places at the expense of slightly less accuracy (approved by beard). Swapper PR to follow removing logic there as well
- Differentiate between "trade" and "quote" and use the correct values in correct locations.
- getQuote flow has been entirely re-written to be far simpler and now correctly debounces
- Lots of removed unnecesary code & renames to make more sense

This PR might be easier to understand by looking at swapper as a whole instead of trying to figure out the diff. a lot has changed.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

https://github.com/shapeshift/web/issues/1689

## Risk

Trading could be borked if I messed something u[p

## Testing

Make sure that all aspects of swapping behave the same as before. Confirm that everything looks correct and loads prices for for:

- Selecting buy asset
- Selecting sell assert
- Toggle sell asset with buy asset
- Go to confirm screen with a token that will need approval and see that it correctly shows approval tx
- Go to confirm screen and see that all values make sense
- Do a trade and see it work

### CODE REVIEW

File: src/components/Approval/Approval.tsx [hunk 1]
```diff
@@ -81,35 +81,8 @@ export const Approval = () => {
...
-      if (!result?.success && result?.statusReason) {
-        handleToast(result.statusReason)
-      }
-
-      if (result?.success) {
-        history.push({ pathname: '/trade/confirm', state: { fiatRate } })
-      } else {
-        history.push('/trade/input')
-      }
+      history.push({ pathname: '/trade/confirm', state: { fiatRate } })
```
Comment (gomesalexandre) [inline src/components/Approval/Approval.tsx:85]: I know this is coming from the previous code, but can we move the routes into an enum while at it?

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx [hunk 2]
```diff
@@ -1,38 +1,31 @@
...
-  sellAsset,
-  buyAsset,
+  tradeFiatAmount,
+  buyIcon,
+  trade,
   boxSize = '24px',
   status,
   ...rest
-}: AssetToAssetProps) => {
+}: AssetToAssetProps<ChainTypes>) => {
```
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/AssetToAsset.tsx:24]: If we're not using a specific `ChainTypes`, might as well default the generic

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx [hunk 3]
```diff
@@ -133,6 +133,12 @@ export const TradeConfirm = ({ history }: RouterProps) => {
     history.push('/trade/input')
   }
 
+  const tradeFiatAmount = toFiat(
+    bnOrZero(fromBaseUnit(bnOrZero(trade?.sellAmount), trade?.sellAsset.precision ?? 0))
```
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:137]: `bnOrZero()` will always do an `unwrap_or_default` style defaulting to 0 for falsy values, no need to use the nullish coercing operator to have a 0 expression on the falsy path

File: src/components/Trade/TradeInput.tsx [hunk 4]
```diff
@@ -48,34 +41,22 @@ export const TradeInput = ({ history }: RouterProps) => {
...
-    TS['quote'],
-    TS['action'],
-    TS['buyAsset'],
-    TS['sellAsset'],
-    TS['estimatedGasFees'],
-  ]
-  const { getQuote, buildQuoteTx, reset, checkApprovalNeeded, getFiatRate, getSendMaxAmount } =
-    useSwapper()
+  const [quote, buyTradeAsset, sellTradeAsset, sellAssetFiatRate] = useWatch({
+    name: ['quote', 'buyAsset', 'sellAsset', 'sellAssetFiatRate'],
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:45]: Ditto, that's coming back from previous code but would live better in an enum

File: src/components/Trade/TradeInput.tsx [hunk 5]
```diff
@@ -84,89 +65,75 @@ export const TradeInput = ({ history }: RouterProps) => {
...
       })
-      fnLogger.trace({ fn: 'buildQuoteTx', result }, 'Quote Completed')
-
-      if (!result?.success && result?.statusReason) {
-        handleToast(result.statusReason)
+      if (!result?.success && result?.statusReason) handleToast(result.statusReason)
+      const approvalNeeded = await checkApprovalNeeded(wallet)
+      if (approvalNeeded) {
+        history.push({
+          pathname: '/trade/approval',
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:92]: Ditto route

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 6]
```diff
@@ -238,189 +230,80 @@ export const useSwapper = () => {
...
-          // TODO: Implement fiatPerUsd here
-          setFees(newQuote, sellAsset)
-          setValue('quote', newQuote)
-          setValue('sellAsset.fiatRate', sellAssetUsdRate.toString())
-          setValue('buyAsset.fiatRate', buyAssetUsdRate.toString())
-          if (action) onFinish(newQuote)
-        } catch (e) {
-          console.error(e)
-        }
+  const updateQuoteDebounced = useRef(
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:239]: What's the rationale for using `useRef` here?

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 7]
```diff
@@ -238,189 +230,80 @@ export const useSwapper = () => {
...
-      }
-    }
-
-    await getQuoteFromSwapper<typeof sellAsset.currency.chain>({
-      amount: formattedAmount,
-      sellAsset: sellAsset.currency,
-      buyAsset: buyAsset.currency,
+    forceQuote,
+  }: GetQuoteInput) => {
+    if (!forceQuote && amount === '0') return
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:286]:
```suggestion
    if (!forceQuote && bnOrZero(amount).isZero()) return
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 8]
```diff
@@ -238,189 +230,80 @@ export const useSwapper = () => {
...
+    setValue('quote', undefined)
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:287]: Ditto field values living in an enum

**Review by gomesalexandre — COMMENTED** [review]
A few early comments, will give this a second pass and test it out

File: src/components/Trade/hooks/useSwapper/useSwapper.ts (hunk 6 above)
Comment (elmutt) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:239]: In order for lodash debounce to work correctly here it needs to use the single reference is my understnaiong. 

Other people online recommend doing it like this but if theres a better way im down

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx (hunk 2 above)
Comment (elmutt) [reply src/components/Trade/TradeConfirm/AssetToAsset.tsx:24]: Not sure how to do that

IMO what I have here works fine but willing to change it if theres a better way

File: src/components/Trade/hooks/useSwapper/calculateAmounts.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    fiatSellAmount = bnOrZero(amount).times(bnOrZero(buyAssetUsdRate)).toFixed(2)
+  } else if (action === TradeAmountInputField.FIAT) {
+    sellAmount = bnOrZero(amount).dividedBy(sellAssetUsdRate)
+    buyAmount = bnOrZero(amount).dividedBy(buyAssetUsdRate)
+    fiatSellAmount = amount
+  } else {
+    sellAmount = '0'
+    buyAmount = '0'
+    fiatSellAmount = '0'
+  }
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/calculateAmounts.ts:55]: can we wrap all of this in a function that takes an action, uses a switch, and returns a tuple of `[sellAmount, buyAmount, fiatSellAmount` in the correct format?

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 10]
```diff
@@ -70,53 +67,72 @@ export const useTradeRoutes = (
...
+        const previousBuyAsset = { ...getValues('buyAsset') }
+
+        // Handle scenario where same asset is selected for buy and sell
+        if (asset.assetId === previousSellAsset?.asset.assetId) {
+          setValue('buyAsset.asset', asset)
+          setValue('sellAsset.asset', previousBuyAsset.asset)
+        } else {
+          setValue('buyAsset.asset', asset)
+          setValue('sellAsset.asset', sellTradeAsset?.asset)
+        }
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:119]: same deal, this is all redundant if we don't give the user the option to select the same asset

File: src/components/Trade/types.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  FIAT = 'FIAT',
+}
+
+export type TradeAsset = {
+  asset: Asset
+  amount?: string
+}
+
+export type TradeProps = {
+  defaultBuyAssetId?: AssetId
```
Comment (0xdef1cafe) [inline src/components/Trade/types.ts:17]:
```suggestion
  defaultBuyAssetId: AssetId
```

shouldn't a default always be defaulted?

File: src/components/Trade/Trade.tsx [hunk 12]
```diff
@@ -1,51 +1,20 @@
...
-  fees?: chainAdapters.QuoteFeeData<C>
-  trade?: MinMax
-  action?: TradeActions
-  fiatAmount?: string
-  quote?: Quote<C>
-  estimatedGasFees?: string
-}
-
 export const Trade = ({ defaultBuyAssetId }: TradeProps) => {
   const methods = useForm<TradeState<ChainTypes>>({
```
[the remark this answers is not in the document]
Comment (elmutt) [reply src/components/Trade/Trade.tsx:14]: yes ryan and I have a parallel PR for lib & web. 

unfortunately it is going to be a breaking change and lib needs to go in first

File: src/components/Trade/hooks/useSwapper/calculateAmounts.ts (hunk 9 above)
Comment (elmutt) [reply src/components/Trade/hooks/useSwapper/calculateAmounts.ts:55]: we can. is that actually better tho?

File: src/components/Approval/Approval.tsx [hunk 13]
```diff
@@ -208,7 +181,7 @@ export const Approval = () => {
                       isExternal
                       color='blue.500'
                       // TODO:(ryankk) create explorer links given a link template and a value
```
[the remark this answers is not in the document]
Comment (elmutt) [reply src/components/Approval/Approval.tsx:183]: dont think so. removing

File: src/components/Trade/hooks/useSwapper/calculateAmounts.ts (hunk 9 above)
Comment (0xdef1cafe) [reply src/components/Trade/hooks/useSwapper/calculateAmounts.ts:55]: then we can do

```
return {
    ...amountsFromAction(action),
    sellAssetUsdRate: sellAssetUsdRate.toString(),
    buyAssetUsdRate: sellAssetUsdRate.toString(),
  }
```

where `amountsFromAction` is pure and testable

File: src/components/Trade/hooks/useSwapper/useSwapper.ts (hunk 6 above)
Comment (gomesalexandre) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:238]: Wouldn't `useCallback` keep the referential stability without having to deal with refs?

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx (hunk 2 above)
Comment (gomesalexandre) [reply src/components/Trade/TradeConfirm/AssetToAsset.tsx:24]:
```typescript
type AssetToAssetProps<C extends ChainTypes = ChainTypes> = {
``` 

Tried this locally and it works, no need to pass a generic

**Review by DaoDev44 — APPROVED** [review]

**Review by cjthompson — APPROVED** [review]

**Pull request merged**